File: packages/sdk/src/sdk/signatures/sign-modify-lock.ts
import type { WalletClient } from 'viem'
import type { Address } from '../types'
import { createDomain, MODIFY_LOCK_TYPES, type ModifyLockMessage } from './eip712-types'

export interface SignModifyLockParams {
  walletClient: WalletClient
  chainId: number
  verifyingContract: Address
  message: ModifyLockMessage
}

export async function signModifyLockMessage({
  walletClient,
  chainId,
  verifyingContract,
  message,
}: SignModifyLockParams): Promise<`0x${string}`> {
  const account = walletClient.account
  if (!account) {
    throw new Error('No account connected to wallet client')
  }

  const domain = createDomain(chainId, verifyingContract)

  const signature = await walletClient.signTypedData({
    account,
    domain,
    types: MODIFY_LOCK_TYPES,
    primaryType: 'ModifyLock',
    message,
  })

  return signature
}

File: sdk/py/src/flexvaults/signatures/sign_modify_lock.py
from __future__ import annotations

from dataclasses import dataclass

from eth_account.signers.local import LocalAccount

from ..types import Address, HexString, Network
from ._signer import sign_typed_data
from .eip712_types import MODIFY_LOCK_TYPES, ModifyLockMessage, create_domain


@dataclass
class SignModifyLockParams:
    account: LocalAccount
    network: Network
    verifying_contract: Address
    message: ModifyLockMessage


def sign_modify_lock_message(params: SignModifyLockParams) -> HexString:
    domain = create_domain(params.network, params.verifying_contract)

    message_data = {
        "lockId": params.message.lock_id,
        "amount": params.message.amount,
        "newExpiry": params.message.new_expiry,
        "nonce": params.message.nonce,
    }

    return sign_typed_data(
        account=params.account,
        domain=domain,
        types=MODIFY_LOCK_TYPES,
        primary_type="ModifyLock",
        message=message_data,
    )
